src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -12402,65 +12365,46 @@ void Unit::SetStunned(bool apply)
             RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_STUNNED);
 
         if (!HasUnitState(UNIT_STATE_ROOT))         // prevent moving if it also has root effect
-        {
-            WorldPacket data(SMSG_FORCE_MOVE_UNROOT, 8+4);
-            data << GetPackGUID();
-            data << uint32(0);
-            SendMessageToSet(&data, true);
-
-            RemoveUnitMovementFlag(MOVEMENTFLAG_ROOT);
-        }
+            SetRooted(false);
     }
 }
 
 void Unit::SetRooted(bool apply)
+{    
+    // do nothing if the unit is already in the required state
+    if ((HasUnitMovementFlag(MOVEMENTFLAG_ROOT) || HasUnitMovementFlag(MOVEMENTFLAG_PENDING_ROOT)) == apply && !HasPendingMovementChange(ROOT))
+        return;
+
+    // prevent moving if it also has stun effect
+    if (!apply && HasUnitState(UNIT_STATE_STUNNED))
```

Comment:
My bad. I think I misunderstood what you said. I thought you were talking about a bug already in 3.3.5, not a regression bug provoked by the changes of this PR. Anyway, I reproduced the bug and indeed these lines were at fault. Thanks